### PR TITLE
feat(android-distribution): add httpclient for checking for build distribution updates

### PR DESCRIPTION
**Pull request by runningcode**
## Summary

Implements the `checkForUpdateBlocking` method in `DistributionIntegration` to check for app updates via Sentry's distribution API.

## Why not reuse existing HttpConnection?

The existing `HttpConnection` class is designed specifically for Sentry event transport and is not suitable for distribution API calls:
- Hardcoded for POST requests (we need GET)
- Expects Sentry envelopes with gzip encoding (we need simple JSON)
- Only considers status 200 successful (REST APIs use 200-299 range)
- Includes Sentry-specific rate limiting logic

## Changes

- **DistributionHttpClient**: New HTTP client for distribution API requests
  - Supports GET requests with query parameters (main_binary_identifier, app_id, platform, version)
  - Uses SentryOptions.DistributionOptions for configuration (orgSlug, projectSlug, orgAuthToken)

- **UpdateResponseParser**: JSON response parser for API responses

#skip-changelog

🤖 Generated with [Claude Code](https://claude.ai/code)

---

> [!NOTE]
> Adds an HTTP client and JSON parser for distribution update checks, implements checkForUpdateBlocking with network/error handling, updates API types, and adds tests.
> 
> - **Android Distribution**:
>   - **Update Check Flow**: Implements `DistributionIntegration.checkForUpdateBlocking` using new `DistributionHttpClient` and `UpdateResponseParser`; builds params from app package info and opens download via `downloadUpdate`.
>   - **Networking & Errors**: Adds detailed error handling (e.g., `UnknownHostException`, `SocketTimeoutException`) and logs; introduces `UpdateStatus.NoNetwork` result.
>   - **API Changes**: `UpdateInfo.getCreatedDate()` becomes nullable; new `UpdateStatus.NoNetwork` class.
> - **Testing**:
>   - Adds unit tests `DistributionHttpClientTest` (ignored real API test) and `UpdateResponseParserTest` (covers success, errors, edge cases).
> - **Build**:
>   - Enables unit test Android resources/default values and adds test dependencies in `build.gradle.kts`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ff0acf5ed2473f50445a52baa24a81f66e3dc9ad. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
+import java.net.HttpURLConnection
+import java.net.URL
+import java.net.URLEncoder
+import javax.net.ssl.HttpsURLConnection
+
+/** HTTP client for making requests to Sentry's distribution API. */
+internal class DistributionHttpClient(private val options: SentryOptions) {
```
Comment (runningcode) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt:14]: I had claude generate an HttpClient for me. It works. I'm just not sure if it is a good idea to develop another client.

The other alternative is to expand the existing `HttpConnection` to support our use case OR shadow oktthp in to our library. Open to thoughts!

File: sentry-android-distribution/src/test/java/io/sentry/android/distribution/DistributionHttpClientTest.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package io.sentry.android.distribution
+
+import io.sentry.SentryOptions
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Ignore
+import org.junit.Test
+
+class DistributionHttpClientTest {
```
Comment (runningcode) [inline sentry-android-distribution/src/test/java/io/sentry/android/distribution/DistributionHttpClientTest.kt:9]: This is just a test class I used to test the integration. In a future, SAGP will inject the authToken and other things so we won't need to specify this manually. I'll also integrate distribution in to the sample app.

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      options.logger.log(SentryLevel.ERROR, e, "Network error while checking for updates")
+      throw e
+    }
+  }
+
+  private fun makeRequest(url: URL, authToken: String): HttpResponse {
+    val connection = url.openConnection() as HttpURLConnection
+
+    try {
+      // Configure connection
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt]: nit: the classic claude comments I don't find valuable normally

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      connection.setRequestProperty("Authorization", "Bearer $authToken")
+      connection.setRequestProperty("Accept", "application/json")
+      connection.setRequestProperty(
+        "User-Agent",
+        options.sentryClientName ?: "sentry-android-distribution",
+      )
+      connection.connectTimeout = options.connectionTimeoutMillis
+      connection.readTimeout = options.readTimeoutMillis
+
+      // Set SSL socket factory if available
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt]: same here and below

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    if (orgSlug.isEmpty() || projectSlug.isEmpty() || authToken.isEmpty()) {
+      throw IllegalStateException(
+        "Missing required distribution configuration: orgSlug, projectSlug, or orgAuthToken"
+      )
+    }
+
+    val queryParams = buildQueryParams(params)
+    val url =
+      URL(
+        "$baseUrl/api/0/projects/$orgSlug/$projectSlug/preprodartifacts/check-for-updates/?$queryParams"
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt]: Could use https://developer.android.com/reference/android/net/Uri.Builder.html here

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+import io.sentry.SentryLevel
+import io.sentry.SentryOptions
+import io.sentry.UpdateInfo
+import io.sentry.UpdateStatus
+import org.json.JSONException
+import org.json.JSONObject
+
+/** Parser for distribution API responses. */
+internal class UpdateResponseParser(private val options: SentryOptions) {
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt:11]: It might be worth having some tests for this also since it is nicely separated

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt [hunk 7]
```diff
@@ -44,7 +54,19 @@ public class DistributionIntegration(context: Context) : Integration, IDistribut
...
+
+      val params = createUpdateCheckParams()
+      val response = httpClient.checkForUpdates(params)
+      responseParser.parseResponse(response.statusCode, response.body)
+    } catch (e: IllegalStateException) {
+      sentryOptions.logger.log(SentryLevel.WARNING, e.message ?: "Configuration error")
+      UpdateStatus.UpdateError(e.message ?: "Configuration error")
+    } catch (e: Exception) {
+      sentryOptions.logger.log(SentryLevel.ERROR, e, "Failed to check for updates")
+      UpdateStatus.UpdateError("Network error: ${e.message}")
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: It might be nice to be able to distinguish between 'not connected' and 'network error' in the return type if that is possible and easy.

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt [hunk 8]
```diff
@@ -75,4 +97,30 @@ public class DistributionIntegration(context: Context) : Integration, IDistribut
...
+          packageManager.getPackageInfo(packageName, PackageManager.PackageInfoFlags.of(0))
+        } else {
+          @Suppress("DEPRECATION") packageManager.getPackageInfo(packageName, 0)
+        }
+
+      val versionName = packageInfo.versionName ?: "unknown"
+      val appId = context.applicationInfo.packageName
+
+      DistributionHttpClient.UpdateCheckParams(
+        mainBinaryIdentifier = appId, // Using package name as binary identifier
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: Do we want this?

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt [hunk 9]
```diff
@@ -75,4 +97,30 @@ public class DistributionIntegration(context: Context) : Integration, IDistribut
...
+        appId = appId,
+        platform = "android",
+        version = versionName,
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: I think we're missing version code iirc (Android has both a human facing version and a machine facing version for apps, one is versionCode and the other versionNumber)

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    val connection = url.openConnection() as HttpURLConnection
+
+    try {
+      // Configure connection
+      connection.requestMethod = "GET"
+      connection.setRequestProperty("Authorization", "Bearer $authToken")
+      connection.setRequestProperty("Accept", "application/json")
+      connection.setRequestProperty(
+        "User-Agent",
+        options.sentryClientName ?: "sentry-android-distribution",
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt]: We maybe should default to the same as here:
https://github.com/getsentry/sentry-java/blob/d21770841d84eddba0636e2082c029db2a158457/sentry-android-core/src/main/java/io/sentry/android/core/SentryAndroidOptions.java#L231

(or, since it seems like it is always set despite being nullable we could throw if it is unset)

**Review by chromy — COMMENTED** [review]
Thanks!

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 8 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: nope, will remove

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 9 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: Yes good point. I didn’t see versionCode as a parameter to the check for updates APi.
https://github.com/getsentry/sentry/blob/master/src/sentry/preprod/api/endpoints/project_preprod_check_for_updates.py#L62

Should I pass this in as versionName + versionCode or should we add an extra parameter to the API?

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt (hunk 10 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionHttpClient.kt]: sounds good, i added a throw if unset

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 7 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: `UnknownHostException` would mean there is possibly no network and `SocketTimeoutException` could mean a bad network. I added those to distinguish here.

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 9 above)
Comment (chromy) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: We should add an extra parameter, we called it 'build' before: https://github.com/EmergeTools/emerge/blob/97c2adfe7e133e871c37beae2ff1b618002b87bd/backend/services/emerge/src/distribution/api/checkForUpdates.ts#L25C9-L25C22.

In Sentry backend it's called `build_number` https://github.com/getsentry/sentry/blob/46a9e4fd4fedb3f195b77db0b137c7daf167e8e2/src/sentry/preprod/models.py#L110C5-L110C17

We can call the parameter `build_number` maybe? That seems more clear then `build` for me.

- `versionName` should be `version`
- `versionCode` should be the other one

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt (hunk 6 above)
Comment (chromy) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt:11]: thanks!

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+  private fun parseUpdateInfo(json: JSONObject): UpdateInfo {
+    val id = json.optString("id", "")
+    val buildVersion = json.optString("buildVersion", "")
+    val buildNumber = json.optInt("buildNumber", 0)
+    val downloadUrl = json.optString("downloadUrl", "")
+    val appName = json.optString("appName", "")
+    val createdDate = json.optString("createdDate", "")
+
+    // Validate required fields (optString returns "null" for null values)
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt:61]: What an odd API :/

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+
+  private fun parseUpdateInfo(json: JSONObject): UpdateInfo {
+    val id = json.optString("id", "")
+    val buildVersion = json.optString("buildVersion", "")
+    val buildNumber = json.optInt("buildNumber", 0)
+    val downloadUrl = json.optString("downloadUrl", "")
+    val appName = json.optString("appName", "")
+    val createdDate = json.optString("createdDate", "")
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt:59]: If it's fine for this to be unset then maybe it should be nullable in the `UpdateInfo` type

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    // Validate required fields (optString returns "null" for null values)
+    if (
+      id.isEmpty() ||
+        id == "null" ||
+        buildVersion.isEmpty() ||
+        buildVersion == "null" ||
+        downloadUrl.isEmpty() ||
+        downloadUrl == "null"
+    ) {
+      throw IllegalArgumentException("Missing required update information in API response")
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt]: Lets try and give a better error message for future us, either could do this more like:

```
if (is.isEmpty()) {
  throw IllegalArgumentException("id must be set")
}
```

or could just do:

```
if (...) {
  throw IllegalArgumentException(Missing required update information in API response '" + body + "'")
}
```
or something

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt [hunk 14]
```diff
@@ -44,7 +56,31 @@ public class DistributionIntegration(context: Context) : Integration, IDistribut
...
+      sentryOptions.logger.log(SentryLevel.WARNING, e.message ?: "Configuration error")
+      UpdateStatus.UpdateError(e.message ?: "Configuration error")
+    } catch (e: UnknownHostException) {
+      // UnknownHostException typically indicates no internet connection available
+      sentryOptions.logger.log(
+        SentryLevel.ERROR,
+        e,
+        "DNS lookup failed - check internet connection",
+      )
+      UpdateStatus.UpdateError("No internet connection or invalid server URL")
```
Comment (chromy) [inline sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: We could add a state for this and below. `UpdateStatus.NoNetwork` or `UpdateStatus.Unknown`

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 9 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: Ah i see, but it looks like the backend API doesn’t support this yet. I’ll look in to adding this. https://github.com/getsentry/sentry/blob/master/src/sentry/preprod/api/endpoints/project_preprod_check_for_updates.py#L65

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 14 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: added!

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt (hunk 13 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/UpdateResponseParser.kt]: added much clearer errors!

File: sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt (hunk 9 above)
Comment (runningcode) [reply sentry-android-distribution/src/main/java/io/sentry/android/distribution/DistributionIntegration.kt]: added
